crates/air_r_formatter/src/string_literal.rs

```diff
@@ -0,0 +1,147 @@
+use air_r_syntax::RSyntaxKind::R_STRING_LITERAL;
+use air_r_syntax::RSyntaxToken;
+use biome_formatter::prelude::syntax_token_cow_slice;
+use biome_formatter::prelude::Formatter;
+use biome_formatter::trivia::format_replaced;
+use biome_formatter::Format;
+use biome_formatter::FormatResult;
+use std::borrow::Cow;
+use std::cell::Cell;
+
+use crate::context::RFormatContext;
+use crate::RFormatter;
+
+/// Helper utility for formatting a string literal token
+///
+/// The main job of this utility is to `normalize()` the string and handle the
+/// complicated way we have to call [format_replaced] with that normalized result.
+pub(crate) struct FormatStringLiteralToken<'token> {
+    /// The string literal token to format
+    token: &'token RSyntaxToken,
+}
+
+impl<'token> FormatStringLiteralToken<'token> {
+    pub(crate) fn new(token: &'token RSyntaxToken) -> Self {
+        Self { token }
+    }
+
+    fn normalize(&self) -> FormatNormalizedStringLiteralToken {
+        let token = self.token;
+
+        debug_assert!(
+            matches!(token.kind(), R_STRING_LITERAL),
+            "Found kind {:?}",
+            token.kind()
+        );
+
+        let text = token.text_trimmed();
+        let text = normalize_string(text);
+        let text = Cell::new(text);
+
+        FormatNormalizedStringLiteralToken { token, text }
+    }
+}
+
+impl Format<RFormatContext> for FormatStringLiteralToken<'_> {
+    fn fmt(&self, f: &mut RFormatter) -> FormatResult<()> {
+        self.normalize().fmt(f)
+    }
+}
+
+struct FormatNormalizedStringLiteralToken<'token> {
+    /// The original string literal token before normalization
+    token: &'token RSyntaxToken,
+
+    /// The normalized text
+    /// Wrapped in a [`Cell`] to avoid having to clone when passing to [`syntax_token_cow_slice`]
+    text: Cell<Cow<'token, str>>,
+}
+
+impl Format<RFormatContext> for FormatNormalizedStringLiteralToken<'_> {
+    fn fmt(&self, f: &mut Formatter<RFormatContext>) -> FormatResult<()> {
+        format_replaced(
+            self.token,
+            &syntax_token_cow_slice(
+                self.text.take(),
+                self.token,
+                self.token.text_trimmed_range().start(),
+            ),
+        )
+        .fmt(f)
+    }
+}
+
+/// Normalize a string, returning a [`Cow::Borrowed`] if the input was already normalized
```

Comment:
Maybe mention why using the `line_ending` won't work here, because it mutates strings in place?

Reply:
I also anticipate a small chance that we might do further "normalization" here, like using a consistent quote style or something similar